File: leetcode/Python/01-matrix.py
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        max_val, wy, wx = 100000000, len(mat), len(mat[0])
        for i in range(wy):
            for j in range(wx):
                if mat[i][j]==0:
                    continue
                mat[i][j] = max_val
                if i>0:
                    mat[i][j]=min(mat[i][j],mat[i-1][j]+1)
                if j>0:
                    mat[i][j]=min(mat[i][j],mat[i][j-1]+1)
        for i in range(wy-1,-1,-1):
            for j in range(wx-1,-1,-1):
                if mat[i][j]==0:
                    continue
                if i+1<wy:
                    mat[i][j]=min(mat[i][j],mat[i+1][j]+1)
                if j+1<wx:
                    mat[i][j]=min(mat[i][j],mat[i][j+1]+1)
        return mat

'''
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        y, x, ny, nx, dy, dx, = 0, 0, 0, 0, [1,-1,0,0], [0,0,1,-1]
        wy, wx, max_val = len(mat), len(mat[0]), len(mat)*len(mat[0])
        dq = collections.deque()
        for i in range(len(mat)):
            for j in range(len(mat[0])):
                if mat[i][j]>0:
                    mat[i][j]=max_val
                else:
                    dq.append([i,j])
        while dq:
            y, x = dq.popleft()
            for i in range(4):
                ny, nx = y+dy[i], x+dx[i]
                if ny<0 or ny>=wy or nx<0 or nx>=wx or mat[y][x]+1 >= mat[ny][nx]:
                    continue
                mat[ny][nx]=mat[y][x]+1
                dq.append([ny,nx])
        return mat
'''

File: leetcode/C++/01-matrix.cpp
class Solution {
public:
    const int MAX = 1000000000;
    vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
        int wy = mat.size(), wx = mat[0].size();
        for(int i=0; i<wy; ++i){
            for(int j=0; j<wx; ++j){
                if(mat[i][j]==0) continue;
                mat[i][j]=MAX;
                if(i) mat[i][j]=min(mat[i][j],mat[i-1][j]+1);
                if(j) mat[i][j]=min(mat[i][j],mat[i][j-1]+1);
            }
        }
        for(int i=wy-1; i>=0; --i){
            for(int j=wx-1; j>=0; --j){
                if(mat[i][j]==0) continue;
                if(i+1<wy) mat[i][j]=min(mat[i][j],mat[i+1][j]+1);
                if(j+1<wx) mat[i][j]=min(mat[i][j],mat[i][j+1]+1);
            }
        }
        return mat;
    }
};

/*
class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
        int MAX = mat.size()*mat[0].size()+1, wy=mat.size(), wx=mat[0].size();
        int y,x,ny,nx, dy[]={1,-1,0,0}, dx[]={0,0,1,-1};
        queue<pair<int,int>> q;
        for(int i=0; i<wy; ++i){
            for(int j=0; j<wx; ++j){
                if(mat[i][j]) mat[i][j]=MAX;
                else q.emplace(i,j);
            }
        }
        while(!q.empty()){
            y=q.front().first; x=q.front().second; q.pop();
            for(int i=0; i<4; ++i){
                ny=y+dy[i]; nx=x+dx[i];
                if(ny<0 || ny>=wy || nx<0 || nx>=wx || mat[y][x] + 1 >= mat[ny][nx]) continue;
                mat[ny][nx]=mat[y][x]+1;
                q.emplace(ny,nx);
            }
        }
        return mat;
    }
};
*/